internal/cms/signed.go

```
@@ -0,0 +1,313 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cms
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/notaryproject/tspclient-go/internal/encoding/asn1/ber"
+	"github.com/notaryproject/tspclient-go/internal/hashutil"
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// ParsedSignedData is a parsed SignedData structure for golang friendly types.
+type ParsedSignedData struct {
+	// Content is the content of the EncapsulatedContentInfo.
+	Content []byte
+
+	// ContentType is the content type of the EncapsulatedContentInfo.
+	ContentType asn1.ObjectIdentifier
+
+	// Certificates is the list of certificates in the SignedData.
+	Certificates []*x509.Certificate
+
+	// CRLs is the list of certificate revocation lists in the SignedData.
+	CRLs []x509.RevocationList
+
+	// SignerInfos is the list of signer information in the SignedData.
+	SignerInfos []SignerInfo
+}
+
+// ParseSignedData parses ASN.1 BER-encoded SignedData structure to golang
+// friendly types.
+func ParseSignedData(berData []byte) (*ParsedSignedData, error) {
+	data, err := ber.ConvertToDER(berData)
+	if err != nil {
+		return nil, SyntaxError{Message: "invalid signed data: failed to convert from BER to DER", Detail: err}
+	}
+	var contentInfo ContentInfo
+	if _, err := asn1.Unmarshal(data, &contentInfo); err != nil {
+		return nil, SyntaxError{Message: "invalid content info: failed to unmarshal DER to ContentInfo", Detail: err}
+	}
+	if !oid.SignedData.Equal(contentInfo.ContentType) {
+		return nil, ErrNotSignedData
+	}
+
+	var signedData SignedData
+	if _, err := asn1.Unmarshal(contentInfo.Content.Bytes, &signedData); err != nil {
+		return nil, SyntaxError{Message: "invalid signed data", Detail: err}
+	}
```

We should validate CMSVersion and fail fast for unsupported.

---

Also validate digestAlgorithms of SignedData are valid values and contain atleast one valid value as per notation?

---

> Also validate digestAlgorithms of SignedData are valid values and contain atleast one valid value as per notation?

@priteshbandi This is done in notation-core-go as it's not a general requirement for cms/timestamp.

---

@priteshbandi added CMSVersion check.